internal/transform/scope-css.go

```diff
@@ -10,52 +11,200 @@ import (
+					case "global(":
+						parenCount++
+						// omit from output
+					case "(":
+						parenCount++
+						out += strVal
+					case ")":
+						parenCount--
+						if !isGlobal || parenCount != 0 {
+							out += strVal // output only if this doesn’t close ":global("
+						}
+					case "[":
+						isBracket = true
+						isElement = false
+						// if there is no selector before an attribute selector, then assume "*"
+						if n == 0 {
+							out += scopeRule("", opts)
 						}
-					} else {
-						out += string(data)
+						out += strVal
+					case "]":
+						isBracket = false
+						out += strVal
+					case "{":
+						keyframeCurlyCount++
+						out += strVal
+					case "}":
+						keyframeCurlyCount--
+						if keyframeCurlyCount == 0 {
+							isKeyframes = false
+						}
+						out += strVal
+					case "*":
+						if !isGlobal {
+							out += scopeRule("", opts) // turns "*" into ".astro-XXXXXX" rather than "*.astro-XXXXXX"
+						} else {
+							out += strVal
+						}
+					default:
+						// handle IDs with parens attached
+						if strings.Contains(strVal, "(") {
+							parenCount++ // if new paren opened, count it
+						}
+
+						// important: :global() might be chained (:global().some-class)
+						// so keep it active until whitespace is reached after final paren
+						if val.TokenType == css.WhitespaceToken && isGlobal && parenCount == 0 {
+							isGlobal = false
+						}
+
+						if !isPseudoState && // don’t scope pseudostates
+							!isGlobal && // don’t scope in :global() scope
+							!isBracket && // don’t scope within element brackets
+							parenCount == 0 && // don’t scope within parens like :not()
+							gt == css.BeginRulesetGrammar && val.TokenType == css.IdentToken && // only scope scope-able CSS properties
+							(!isElement || (isElement && shouldScope(strVal))) { // don’t scope body, html, etc.
+							out += scopeRule(strVal, opts)
+						} else {
+							out += strVal
+						}
+						isElement = true      // reset element state
+						isPseudoState = false // reset pseudostate status
 					}
 				}
+
+				// generate tail of next rule
+				switch gt {
+				case css.BeginAtRuleGrammar,
+					css.BeginRulesetGrammar:
+					out += "{"
+				case css.DeclarationGrammar,
+					css.EndRulesetGrammar,
+					css.EndAtRuleGrammar:
+					out += ";"
+				case css.QualifiedRuleGrammar:
+					out += ","
+				}
+			default:
+				out += string(data)
+				for _, val := range p.Values() {
+					strVal := string(val.Data)
+					out += strVal
+				}
+				out += ";"
 			}
-			n.FirstChild.Data = out
 		}
+		n.FirstChild.Data = out
 	}
 }
 
-func scopeRule(rule string, opts TransformOptions) string {
-	scope := ".astro-" + opts.Scope
-	return rule + scope
+// Turn ".foo" into ".foo.astro-XXXXXX"
+func scopeRule(id string, opts TransformOptions) string {
+	return id + ".astro-" + opts.Scope
+}
+
+// Get list of elements that should be scoped
+func shouldScope(id string) bool {
+	switch id {
+	case "base",
+		"body",
```

Comment:
`body` seems valid, no?

Reply:
That would turn `body.theme-dark` into `body.theme-dark.astro-XXXXXX` which is what you don’t want, because `<body>` won’t be scoped within a component (there can only be one). It will still scope `body .localclass` properly.